Add Joel's spreadsheet <> dwh blog post (#2444)
fishtown-analytics/docs.getdbt.com: website/blog/2022-11-22-move-spreadsheets-to-your-dwh.md

---
title: "How to move data from spreadsheets into your data warehouse"

> description: "A thankless, humble, and inevitable task: getting spreadsheet data into your data warehouse. In this Developer Blog post, Joel walks through the different options for getting spreadsheet data into your data warehouse, and the pros and cons of each option."

Reviewer: ```suggestion
description: "A thankless, humble, and inevitable task: getting spreadsheet data into your data warehouse. Let's look at some of the different options, and the pros and cons of each."
```

Spreadsheets are the Swiss army knife of data processing. They can add extra context to otherwise inscrutable application identifiers, be the only source of truth for bespoke processes from other divisions of the business, or act as the translation layer between two otherwise incompatible tools.

Because of spreadsheets’ importance as the glue between many business processes, there are different tools to load them into your data warehouse and each one has its own pros and cons, depending on your specific use case.

<!--truncate-->

In general, there are a few questions to ask yourself about your data before choosing one of these tools:

- Who at your company will be loading the data?
- Does it have a consistent format?
- How frequently will it change?
- How big is the dataset?
- Do changes need to be tracked?
- Where are the files coming from?

Let’s have a look at some of the offerings to help you get your spreadsheets into your data warehouse.

## dbt seeds

> If you’re using [dbt](https://www.getdbt.com/) to manage your data transformations, it comes with an inbuilt csv loader ([seeds](https://docs.getdbt.com/docs/building-a-dbt-project/seeds)) to populate your data warehouse with any files you put inside of your project’s `seeds` folder. It will automatically infer data types from your file’s contents, but you can always override it by [providing explicit instructions in your dbt_project.yml](https://docs.getdbt.com/reference/resource-configs/column_types) file.

Reviewer: Two options here: 
- Just remove the link to the dbt homepage (do we really need to link to ourselves?)
- Assume that basically everyone reading this uses dbt
```suggestion
dbt comes with an inbuilt csv loader ([seeds](https://docs.getdbt.com/docs/building-a-dbt-project/seeds)) to populate your data warehouse with any files you put inside of your project’s `seeds` folder. It will automatically infer data types from your file’s contents, but you can always override it by [providing explicit instructions in your dbt_project.yml](https://docs.getdbt.com/reference/resource-configs/column_types) file.
```

> However, since dbt creates these tables by inserting rows one at a time, it doesn’t perform well at scale (there’s no hard limit but syncing a hundred rows should be fine, as opposed to a few thousand). [The dbt docs](https://docs.getdbt.com/docs/building-a-dbt-project/seeds#faqs) suggest using seeds for “files that contain business-specific logic, for example, a list of country codes or user IDs of employees.”

Reviewer: ```suggestion
However, since dbt creates these tables by inserting rows one at a time, it doesn’t perform well at scale (there’s no hard limit but aim for hundreds of rows rather than thousands). [The dbt docs](https://docs.getdbt.com/docs/building-a-dbt-project/seeds#faqs) suggest using seeds for “files that contain business-specific logic, for example, a list of country codes or user IDs of employees.”
```

A big benefit of using seeds is that your file will be checked into source control, allowing you to easily see when the file was updated and retrieve deleted data if necessary.

#### Good fit for:

- Small files such as mapping employee identifiers to employees
- Infrequently modified files such as mapping country codes to country names
- Data that would benefit from source control
- Programmatic control of data types

#### Not a good fit for:

- Files greater than 1MB in size
- Files that need regular updates

## ETL tools

An obvious choice if you have data to load into your warehouse would be your existing [ETL tool](https://www.getdbt.com/analytics-engineering/etl-tools-a-love-letter/) such as Fivetran or Stitch, which we'll dive into in this section. Below is a summary table highlighting the core benefits and drawbacks of certain ETL tooling options for getting spreadsheet data in your data warehouse.

### Summary table

> | Option/connector | Good for frequently updated data | Good for change tracking | Configurable data types |
> | --- | --- | --- | --- |
> | dbt seeds | ❌ | ✅ | ✅ |
> | Fivetran Browser Upload | ✅ | ❌ | ✅ |
> | Fivetran Google Sheets connector | ✅ | ❌ | ❌ |
> | Fivetran Google Drive connector | ❌ | ❌ | ❌ |
> | Stitch Google Sheets integration | ✅ | ❌ | ❌ |
> | Airbyte Google Sheets connector | ✅ | ❌ | ❌ |

Reviewer: I know you thought extra columns might make it clunky, but I think these are important 😅 
```suggestion
| Option/connector | Data updatable after load |  Configurable data types | Multiple tables per schema | Good for large datasets |
| --- | --- | --- | --- | --- |
| dbt seeds | ✅ | ✅ | ✅ | ❌ |
| Fivetran Browser Upload | ✅ | ✅ | ✅ | ✅ |
| Fivetran Google Sheets connector | ✅ | ❌ | ❌ | ✅ |
| Fivetran Google Drive connector | ❌ | ❌ | ✅ | ✅ |
| Stitch Google Sheets integration | ✅ | ❌ | ❌ | ✅ |
| Airbyte Google Sheets connector | ✅ | ❌ | ❌ | ✅ |
```

It now looks like 
![image](https://user-images.githubusercontent.com/7335046/203656662-9ed79acb-11b2-4720-b5ae-4ca9cac908d0.png)

However, keep in mind that there is no source control on these changes or a to revert them; you might want to consider [snapshotting changes](https://docs.getdbt.com/docs/building-a-dbt-project/snapshots) in dbt if that’s a concern.

Also, Fivetran won’t delete records once they’re created, so the only way to remove records created using this process is by manually [deleting](https://docs.getdbt.com/terms/dml#delete) them from your warehouse. If you have an ad-hoc connector, consider having an automated process to drop these tables regularly, especially if you have PII management concerns.

#### Good fit for:

- Files that are frequently updated by someone
- Allowing anyone in the company to upload files
- Ad-hoc data loads
- Updating a table instead of creating a new one
- Basic data type changes (including handling currency columns)
- Larger files

#### Not a good fit for:

- Tracking changes to data
- Complex type mappings

### Fivetran Google Sheets connector

> The main benefit of connecting to Google Sheets instead of a static spreadsheet should be obvious—teammates all over the world can change the sheet and new records will be loaded into your warehouse automatically. [Fivetran’s Google Sheets connector](https://fivetran.com/docs/files/google-sheets) requires some additional implementation details, but collaborative editing can make the additional configuration effort worthwhile.

Reviewer: ```suggestion
The main benefit of connecting to Google Sheets instead of a static spreadsheet should be obvious—teammates can change the sheet from anywhere and new records will be loaded into your warehouse automatically. [Fivetran’s Google Sheets connector](https://fivetran.com/docs/files/google-sheets) requires some additional initial configuration, but collaborative editing can make the effort worthwhile.
```

- Tracking changes to data
- Documents with many tabs

### Fivetran Google Drive connector

I’m a big fan of [Fivetran’s Google Drive connector](https://fivetran.com/docs/files/google-drive); in the past I’ve used it to streamline a lot of weekly reporting. It allows stakeholders to use a tool they’re already familiar with (Google Drive) instead of dealing with another set of credentials. Every file uploaded into a specific folder on Drive (or [Box, or consumer Dropbox](https://fivetran.com/docs/files/magic-folder)) turns into a table in your warehouse.

<Lightbox src="/img/blog/2022-11-22-move-spreadsheets-to-your-dwh/google-drive-uploader.png" title="Fivetran will add each of these csv files to a single schema in your warehouse, making it ideal for regular uploads" />

Like the Google Sheets connector, the data types of the columns are determined automatically. Dates, in particular, are finicky though—if you can control your input data, try to get it into [ISO 8601 format](https://xkcd.com/1179/) to minimize the amount of cleanup you have to do on the other side.

> We use two macros in the dbt_utils package ([get_relations_by_pattern](https://github.com/dbt-labs/dbt-utils#get_relations_by_pattern-source) and [union_relations](https://github.com/dbt-labs/dbt-utils#union_relations-source)) to combine weekly exports from other tools into a single [model](https://docs.getdbt.com/docs/building-a-dbt-project/building-models) for easy cleanup in a staging model. Make sure you grant your transformer account permission to access all tables in the schema (including future ones) to avoid having to manually intervene after every new file is uploaded.

Reviewer: ```suggestion
I used two macros in the dbt_utils package ([get_relations_by_pattern](https://github.com/dbt-labs/dbt-utils#get_relations_by_pattern-source) and [union_relations](https://github.com/dbt-labs/dbt-utils#union_relations-source)) to combine weekly exports from other tools into a single [model](https://docs.getdbt.com/docs/building-a-dbt-project/building-models) for easy cleanup in a staging model. Make sure you grant your transformer account permission to access all tables in the schema (including future ones) to avoid having to manually intervene after every new file is uploaded.
```

- Teams that may be on a budget

#### Not a good fit for:

- Non-string type data you want preserved in your raw source tables in your data warehouse

## Native warehouse integrations

Each of the major data warehouses also has native integrations to import spreadsheet data. While the fundamentals are the same, there are some differences amongst the various warehousing vendors.

### Snowflake

Snowflake’s offerings are the most robust and user-friendly, offering both a [web-based loader](https://docs.snowflake.com/en/user-guide/data-load-web-ui.html) as well as [a bulk importer](https://docs.snowflake.com/en/user-guide/data-load-bulk.html). The web loader is suitable for small to medium files (up to 50MB) and can be used for specific files, all files in a folder, or files in a folder that match a given pattern. It’s also the most provider-agnostic, with support for Amazon S3, Google Cloud Storage, Azure and the local file system.

<Lightbox src="/img/blog/2022-11-22-move-spreadsheets-to-your-dwh/snowflake-uploader.png" title="Snowflake’s web-based Load Data Wizard via the Snowflake Blog https://www.snowflake.com/blog/tech-tip-getting-data-snowflake/" />

### BigQuery

> BigQuery only supports importing data from external sources hosted by Google such as Google Drive and Google Cloud Storage (as BigQuery and Sheets are both Google products, BigQuery is the only platform on this list that has a native integration that doesn't require 3rd-party tooling). The data it references isn’t copied into BigQuery but can be referenced in queries as though it was. If needed, you can write a copy to BigQuery or just leave it as an external source. The team at supercooldata has written [a great primer on using Google Sheets with BigQuery](https://blog.supercooldata.com/working-with-sheets-in-bigquery/).

Reviewer: ```suggestion
BigQuery only supports importing data from external sources hosted by Google such as Google Drive and Google Cloud Storage (as BigQuery and Sheets are both Google products, BigQuery is the only platform on this list that has a native integration that doesn't require 3rd-party tooling). The data it references isn’t copied into BigQuery but can be referenced in queries as though it was. If needed, you can write a copy to BigQuery or just leave it as an external source. The team at supercooldata has written [a great how-to guide on setting up Google Sheets with BigQuery](https://blog.supercooldata.com/working-with-sheets-in-bigquery/).
```

### Redshift

Unsurprisingly for an AWS product, Redshift prefers to [import CSV files from S3](https://docs.aws.amazon.com/redshift/latest/dg/tutorial-loading-data.html). As with Snowflake, this is achieved with the COPY command, and you can easily control which file(s) are imported from the source bucket. Using S3 as a source compared to a web-based loader or Google Drive means this option isn’t as user-friendly for non-technical folks, but is still a great option to sync files that are automatically generated from other tools.

## Conclusion

> Beyond the options we’ve already covered, there’s an entire world of other tools that can load data from your spreadsheets into your data warehouse.

Reviewer: ```suggestion
Beyond the options we’ve already covered, there’s an entire world of other tools that can load data from your spreadsheets into your data warehouse. This is a living document, so if your preferred method isn't listed then please open a PR and I'll check it out.
```

Author: Updated this + linked to docs repo
